### ISSUE TITLE
Release 4.0.4 docs

### ISSUE THREAD

**Issue by edwardchalstrey1** [alan-turing-institute/data-safe-haven-team#21]

**Comment by jemrobinson** [comment]
We need to check whether an SHM and SRE can be deployed from the current release-v4.0.4 (79715276fb5368aa95a4238625a44e8193fe349b)

**Comment by edwardchalstrey1** [comment]
> We need to check whether an SHM and SRE can be deployed from the current release-v4.0.4 (79715276fb5368aa95a4238625a44e8193fe349b)

@jemrobinson I've updated the todo list to complete this issue above, feel free to add any other steps you think are required

**Comment by edwardchalstrey1** [comment]
I'll actually do this before #1467 - this way I can use the new SHM I'll have set up to test changes to the Firewall rule

**Comment by edwardchalstrey1** [comment]
Note: after deploying the SHM I was unable to [configure NPS logging](https://data-safe-haven.readthedocs.io/en/develop/deployment/deploy_shm.html#configure-nps-logging) because the "Network Policy Server" seemed to not exist on DC1

**Comment by edwardchalstrey1** [comment]
After deploying the SHM and running through the setup steps I got to [Configure MFA](https://data-safe-haven.readthedocs.io/en/develop/deployment/deploy_shm.html#configure-mfa) and then got this error:

```
PS C:\Users\domainorangeadmin> & "C:\Program Files\Microsoft\AzureMfa\Config\AzureMfaNpsExtnConfigSetup.ps1"
VERBOSE: Using the provider 'PowerShellGet' for searching packages.
VERBOSE: The -Repository parameter was not specified.  PowerShellGet will use all of the registered repositories.
VERBOSE: Getting the provider object for the PackageManagement Provider 'NuGet'.
VERBOSE: The specified Location is 'https://www.powershellgallery.com/api/v2' and PackageManagementProvider is 'NuGet'.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='MSOnline'' for ''.
VERBOSE: Total package yield:'1' for the specified package 'MSOnline'.
VERBOSE: Skipping installed module MSOnline 1.1.183.66.
Connecting to Microsoft Azure.  Please sign on as a tenant administrator.
Connect-MsolService : Authentication Error: Unexpected authentication failure.
At C:\Program Files\Microsoft\AzureMfa\Config\AzureMfaNpsExtnConfigSetup.ps1:13 char:1
+ Connect-MsolService -Verbose -ErrorAction Stop
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (:) [Connect-MsolService], Exception
    + FullyQualifiedErrorId : System.Exception,Microsoft.Online.Administration.Automation.ConnectMsolService
```

**Comment by jemrobinson** [comment]
Are you running this command on the NPS server?

**Comment by edwardchalstrey1** [comment]
> Are you running this command on the NPS server?

Yes

**Comment by jemrobinson** [comment]
Did you get a prompt (in a web browser) for login credentials?

**Comment by edwardchalstrey1** [comment]
Yes, I logged in with the aad.admin account I set up for this AAD

**Comment by jemrobinson** [comment]
OK, if you run the script again (following the troubleshooting steps) does it still fail?

**Comment by edwardchalstrey1** [comment]
Have tried a couple of times, but this error seems to be slightly different from the ones in the troubleshooting steps, will have a look at those in case it's the same issue though

**Comment by edwardchalstrey1** [comment]
Ah, hadn't set up MFA for the aad.admin account apparently - sorted now

**Comment by edwardchalstrey1** [comment]
@craddm @jemrobinson I've deployed a new test SHM with 4.0.4 called "Orange" and will test building an SRE once I have my new laptop set up. What should we do with regard to the other SHMs set up:
1. Should we update `prod4` to bring it in line with the latest version? Specifically I'm thinking of the [servicebus firewall change](https://github.com/alan-turing-institute/data-safe-haven/pull/1485)
2. Same question for `blue` (or alternatively should we tear down `blue` and use `orange` instead for future testing - I'm not sure if any other significant SHM changes have been made?)
    - Currently I see 3 SREs on `blue` called `satret1`, `satret2` and `t3guac`. Is it worth managing these in the same way on the TRESA [SRE board](https://github.com/orgs/alan-turing-institute/projects/25/views/1) or is that overkill for test SREs?

**Comment by edwardchalstrey1** [comment]
Also, are there any particular tests that should be run other than the SRE smoke tests?

**Comment by jemrobinson** [comment]
1. We can (and should) update the SHM to the latest released version as-and-when this is available. Nothing we use in production should ever be running anything except a numbered release

2. Dev environments can be in whatever broken/working state is appropriate. We should be clear about who's using which ones though and tidy up any that are not being used.

**Comment by craddm** [comment]
For `blue`, I made the service bus firewall change manually when testing if it worked, so we can leave it be.

**Comment by jemrobinson** [comment]
This is because the test tries to run `import <package_name>` for each package. `pip-tools` isn't an importable package, it's just some scripts. See the example for `repro-catalogue` where we check for an executable instead, and try checking for `pip-compile`.

**Comment by edwardchalstrey1** [comment]
@jemrobinson We should now be able to make a release for the `release-v4.0.4` branch - should I just go ahead and do that? afterwards I can PR this branch into `develop` and figure out any conflicts

**Comment by jemrobinson** [comment]
Have you done all the pre-release check? Not sure we've got these written down anywhere (@JimMadge do you know?) but see e.g. [release 4.0.0](https://github.com/alan-turing-institute/data-safe-haven/releases/tag/v4.0.0). We need:

- SHM build log
- SRD build log
- SRE build logs for
    - Guacamole and MSRDS
    - Tier 2 and Tier 3
    - These can be combined (eg. t2guac and t3msrds)
- Filled out security checklist

**Comment by edwardchalstrey1** [comment]
I have the:

- SHM build log
- SRD build log
- SRE Tier 0 build log

I think we should formally decide then which checks should be done for which releases, were all of those things done for each of the previous patch releases e.g. 4.0.3/4.0.2? I would guess no? https://github.com/alan-turing-institute/data-safe-haven/releases

Reading through the [security checklist](https://data-safe-haven.readthedocs.io/en/latest/deployment/security_checklist.html) just now it felt like if anything this is something we should be doing on a per deployment basis - though maybe we think that's too laborious and so we assume that reproducible IAC principles mean we can do it for a version? Even in this case, should we do it for every patch release or just minor or major releases?

**Comment by jemrobinson** [comment]
I agree that these weren't done for previous patch releases, but I'm not sure whether this is still a patch release (bug fixes only) or if it's a minor release now. 

Here's the diff (https://github.com/alan-turing-institute/data-safe-haven/compare/v4.0.3...release-v4.0.4) what do you think @edwardchalstrey1 @JimMadge @craddm ?

**Comment by edwardchalstrey1** [comment]
> Can you post a changelog here @edwardchalstrey1 (in the style of previous release changelogs) so we can judge better?

Looks like @JimMadge made one already: https://github.com/alan-turing-institute/data-safe-haven/releases/tag/untagged-a52657d965ec43f09888

I've updated that with the recent [PRs to the release branch](https://github.com/alan-turing-institute/data-safe-haven/pulls?q=is%3Apr+is%3Aclosed+base%3Arelease-v4.0.4) and dropped "beta" from the release name

**Comment by edwardchalstrey1** [comment]
My thoughts are we should decide and document the checks that we want to do for the 3 levels of versioning, then decide what this release should be and if needed change the number (e.g. 4.1.0) - I'm not against doing the full set of checks you describe @jemrobinson even for patch releases, but I think we should decide that first.

**Comment by jemrobinson** [comment]
Looking at what we've done in the past, it's:

- MAJOR: SHM log, SRE logs covering Guacamole/MSRDS and tier 2/3, SRD log, security checklist, penetration test
- MINOR: SHM log, SRE logs covering Guacamole/MSRDS and tier 2/3, SRD log, security checklist, penetration test (sometimes)
- PATCH: None

What sounds reasonable to you: @edwardchalstrey1 @JimMadge @craddm for each of these release types?

**Comment by edwardchalstrey1** [comment]
What does the penetration test involve?
Where should the logs be saved?

Maybe releases that have security fixes (as this does), should be at least a MINOR and subject to all these tests, whereas releases that only have feature additions, non-security related bug fixes and documentation changes can be patches with no testing except perhaps the usual smoke tests on a deployed SRE?

Also just thinking now, would it be normal in software to bundle unrelated changes into a release in the way we have been doing? This does seem fine to me, but if we're taking this approach as oppose to doing releases for a set of related features, should we have a pre-defined interval, e.g. make a release once per quarter, and decide the patch/minor/major increment based on whatever is in it?

**Comment by jemrobinson** [comment]
The penetration tests are carried out by a third-party (iSTORM). We set up an SHM and a couple of SREs and give them user accounts and various admin accounts. They try to find exploits and report back to us. Typically takes around a week or two weeks for an in-depth test. Results of previous penetration tests are in Sharepoint under `Security assessments > External penetration tests`.

> Maybe releases that have security fixes (as this does), should be at least a MINOR and subject to all these tests 

We've already said that MINOR means "Adds new functionality, but new SREs can be deployed into existing environments" while PATCH means "Fixes bugs only and new SREs can be deployed into existing environments". I think security fixes can still be a PATCH release.

> whereas releases that only have feature additions, non-security related bug fixes and documentation changes can be patches with no testing except perhaps the usual smoke tests on a deployed SRE?

I guess the real question is: what kind of change can we make where we're ~100% certain that it won't break our SHM or SRE scripts. Documentation changes are in that category, removing MSRDS support is not. Where's the dividing line and is it at the current patch/minor boundary?

> Also just thinking now, would it be normal in software to bundle unrelated changes into a release in the way we have been doing? This does seem fine to me, but if we're taking this approach as oppose to doing releases for a set of related features, should we have a pre-defined interval, e.g. make a release once per quarter, and decide the patch/minor/major increment based on whatever is in it?

For security fixes, it's important to get a patch release out as soon as possible. Major and minor don't have that constraint. The problem is that preparing a release takes away from development time so we generally want to have bundled a few important changes to make it worth doing.

**Comment by edwardchalstrey1** [comment]
> Meta point, why is this issue not in the DSH repo?

I think can move it there, in fact, might it be worth having a GH issue template for releases that has a checklist of the above

**Comment by jemrobinson** [comment]
Let's leave this issue here, but create a new GH issue template in the DSH repo with a checklist of whatever we agree here.

**Comment by edwardchalstrey1** [comment]
> PATCH:
> - SRD log
> - SRE log for at least one SRE
> - Security checklist

Although note that the security checklist requires setting up two SREs (t2 and t3)

**Comment by JimMadge** [comment]
Is the argument that if we make a security fix to the SHM, that will always be a major release, so we don't need to check SHM deployment?

**Comment by edwardchalstrey1** [comment]
@jemrobinson question - for the SHM/SRE/SRD logs, where do/should we save these?

**Comment by jemrobinson** [comment]
Save them in Sharepoint in under `release_logs`. Copy the `templates` folder to your new release name and fill in the templates accordingly. You can use `markdown2pdf.sh` to build the PDF versions that get attached to the release.

**Comment by edwardchalstrey1** [comment]
> Save them in Sharepoint in under `release_logs`. Copy the `templates` folder to your new release name and fill in the templates accordingly. You can use `markdown2pdf.sh` to build the PDF versions that get attached to the release.

thanks, I guess we should record that somewhere internally, maybe on the `data-safe-haven-team` README?

Here's a PR for release template: https://github.com/alan-turing-institute/data-safe-haven/pull/1543

**Comment by edwardchalstrey1** [comment]
Closing this issue with the opening of https://github.com/alan-turing-institute/data-safe-haven/issues/1544
